common/src/ws_rpc_client.rs

```diff
@@ -0,0 +1,251 @@
+                self.send_extrinsic_and_wait_until_in_block(jsonreq, result_in)?;
+                let res = result_out.recv()?;
+                info!("inBlock: {}", res);
+                Ok(Some(Hash::from_hex(res)?))
+            }
+            XtStatus::Broadcast => {
+                self.send_extrinsic_and_wait_until_broadcast(jsonreq, result_in)?;
+                let res = result_out.recv()?;
+                info!("broadcast: {}", res);
+                Ok(None)
+            }
+            XtStatus::Ready => {
+                self.send_extrinsic_until_ready(jsonreq, result_in)?;
+                let res = result_out.recv()?;
+                info!("ready: {}", res);
+                Ok(None)
+            }
+            XtStatus::SubmitOnly => {
+                self.send_extrinsic(jsonreq, result_in)?;
+                let res = result_out.recv()?;
+                info!("submitted xt: {}", res);
+                Ok(None)
+            }
+            _ => Err(ApiClientError::UnsupportedXtStatus(exit_on)),
+        };
+
+        // reset the RpcClient handler used by the WebSocket's thread
+        *self
+            .next_handler
+            .lock()
+            .expect("unable to acquire a lock on RpcClient") = None;
+        result
+    }
+}
+
+impl WsRpcClient {
+    fn get(&self, json_req: String, result_in: ThreadOut<String>) -> WsResult<()> {
+        self.send_rpc_request(json_req, result_in, on_get_request_msg)
+    }
+
+    fn send_extrinsic(&self, json_req: String, result_in: ThreadOut<String>) -> WsResult<()> {
+        self.send_rpc_request(json_req, result_in, on_extrinsic_msg_submit_only)
+    }
+
+    fn send_extrinsic_until_ready(
+        &self,
+        json_req: String,
+        result_in: ThreadOut<String>,
+    ) -> WsResult<()> {
+        self.send_rpc_request(json_req, result_in, on_extrinsic_msg_until_ready)
+    }
+
+    fn send_extrinsic_and_wait_until_broadcast(
+        &self,
+        json_req: String,
+        result_in: ThreadOut<String>,
+    ) -> WsResult<()> {
+        self.send_rpc_request(json_req, result_in, on_extrinsic_msg_until_broadcast)
+    }
+
+    fn send_extrinsic_and_wait_until_in_block(
+        &self,
+        json_req: String,
+        result_in: ThreadOut<String>,
+    ) -> WsResult<()> {
+        self.send_rpc_request(json_req, result_in, on_extrinsic_msg_until_in_block)
+    }
+
+    fn send_extrinsic_and_wait_until_finalized(
+        &self,
+        json_req: String,
+        result_in: ThreadOut<String>,
+    ) -> WsResult<()> {
+        self.send_rpc_request(json_req, result_in, on_extrinsic_msg_until_finalized)
+    }
+
+    fn send_rpc_request(
+        &self,
+        jsonreq: String,
+        result_in: ThreadOut<String>,
+        on_message_fn: OnMessageFn,
+    ) -> WsResult<()> {
+        // 1 used by `on_open` of RpcClient + 1 for `close`
+        const MAGIC_SEND_CONST: usize = 2;
+        let (ws_tx, _ws_rx) = mio::channel::sync_channel(MAGIC_SEND_CONST);
```

Comment:
`channel` is deprecated, but you can use `mio-extras`

Reply:
I don't think I can - mismatched types. `Secret` constructor ;) https://docs.rs/ws/0.2.1/src/ws/.cargo/registry/src/github.com-1ecc6299db9ec823/ws-0.2.1/src/communication.rs.html#52